### PR TITLE
Refactor remaining statistics function

### PR DESCRIPTION
**Pull request by vraghavan78**
This is continuation of the efforts done in https://github.com/greenplum-db/gporca/pull/353

Here is what this PR tries to achieve:
* Move logic to compute statistics objects for UNION ALL, PROJECT, SELECT, GROUP BY and a LIMIT into a separate class that has static functions that do the work. This remove the logic from `CStatistics`. The final goal is to kill `IStatistics` and only have `CStatistics` that is a container for statistical information such histograms, width, fempty, etc.
* Move the unittests for statistics on select predicates into its own test
* Get rid of CHistogramUtils and put these function in the CHistogram class
* Clean up some comments and remove dead code

Please look at individual commits if you feel the PR is too big.

### CODE REVIEW

**Comment by vraghavan78** [comment]
@hsyuan we could, but people felt the file was too big to comprehend.

**Comment by hsyuan** [comment]
make sense.

File: libnaucrates/include/naucrates/statistics/CProjectStatsProcessor.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#define GPNAUCRATES_CProjectStatsProcessor_H
+
+namespace gpnaucrates
+{
+
+	class CProjectStatsProcessor
+	{
+		public:
+
+		// limit
```
Comment (bhuvnesh2703) [inline libnaucrates/include/naucrates/statistics/CProjectStatsProcessor.h]: not limit..

File: libnaucrates/include/naucrates/statistics/CGroupByStatsProcessor.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#define GPNAUCRATES_CGroupByStatsProcessor_H
+
+namespace gpnaucrates
+{
+
+	class CGroupByStatsProcessor
+	{
+		public:
+
+		// limit
```
Comment (bhuvnesh2703) [inline libnaucrates/include/naucrates/statistics/CGroupByStatsProcessor.h]: not limit

File: libnaucrates/src/statistics/CUnionAllStatsProcessor.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright 2018 Pivotal, Inc.
+//
+//	@filename:
+//		CUnionAllStatsProcessor.cpp
+//
+//	@doc:
+//		Statistics helper routines for processing project operations
```
Comment (bhuvnesh2703) [inline libnaucrates/src/statistics/CUnionAllStatsProcessor.cpp]: not project

File: libnaucrates/include/naucrates/statistics/CProjectStatsProcessor.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2018 Pivotal, Inc.
+//
+//	@filename:
+//		CProjectStatsProcessor.h
+//
+//	@doc:
+//		Compute statistics for group by operation
```
Comment (bhuvnesh2703) [inline libnaucrates/include/naucrates/statistics/CProjectStatsProcessor.h]: not group by

File: server/include/unittest/dxl/statistics/CFilterCardinalityTest.h [hunk 5]
```diff
@@ -0,0 +1,221 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2017 Pivotal, Inc.
+//
+//	@filename:
+//		CFilterCardinalityTest.h
+//
+//	@doc:
+//		Test for join cardinality estimation
```
Comment (bhuvnesh2703) [inline server/include/unittest/dxl/statistics/CFilterCardinalityTest.h]: join cardinality?

**Pull request merged**